# Polynomial Regression Study

This notebook investigates how increasing polynomial complexity affects:

- Training error
- Validation error
- Overfitting
- Generalization performance

We experiment with polynomial degrees from 1 to 12 using PM2.5 prediction data.

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import PolynomialFeatures # used to generate polynomial and interaction features from existing input features.
from sklearn.preprocessing import StandardScaler

from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

from sklearn.model_selection import cross_val_score

In [4]:
# Load datasets
train_df = pd.read_csv("../data/processed/train.csv")
val_df   = pd.read_csv("../data/processed/val.csv")
test_df  = pd.read_csv("../data/processed/test.csv")

In [5]:
TARGET = "pm25"

X_train = train_df.drop(columns=[TARGET])
y_train = train_df[TARGET]

X_val = val_df.drop(columns=[TARGET])
y_val = val_df[TARGET]

X_test = test_df.drop(columns=[TARGET])
y_test = test_df[TARGET]

In [6]:
X_train.head()

,co,no2,nox,pm10,relativehumidity,temperature
0,2.84,168.1,0.1449,276.0,87.0,8.5
1,4.13,85.4,0.0499,339.0,88.0,13.3
2,2.86,63.7,0.1407,437.0,92.0,9.8
3,0.21,69.6,0.0386,271.0,51.0,15.1
4,2.12,205.9,0.1507,268.0,87.0,11.0


In [7]:
y_train.head()

0    149.0
1    229.0
2    285.0
3    101.0
4    122.0
Name: pm25, dtype: float64

In [10]:
print(X_train.shape)
print(X_val.shape)
print(X_test.shape)

(814, 6)
(162, 6)
(110, 6)


##  Baseline Linear Regression Pipeline

In [11]:
baseline_model = Pipeline([
    ("scaler", StandardScaler()),
    ("model", LinearRegression())
])

baseline_model.fit(X_train, y_train)

pred_train = baseline_model.predict(X_train)
pred_val   = baseline_model.predict(X_val)

train_rmse = np.sqrt(mean_squared_error(y_train, pred_train))
val_rmse   = np.sqrt(mean_squared_error(y_val, pred_val))

print("Train RMSE:", train_rmse)
print("Validation RMSE:", val_rmse)

Train RMSE: 20.39552847106202
Validation RMSE: 17.707039101919356


## Baseline Linear Regression Experiment

As an initial benchmark, a standard Linear Regression model was trained using the preprocessed environmental features.

### Pipeline Components
- **StandardScaler** → Normalizes feature distributions
- **LinearRegression** → Fits a linear relationship between features and PM2.5 concentration

### Objective
The purpose of this experiment is to establish a baseline performance before introducing polynomial feature expansion and regularization techniques.

### Results

| Metric | RMSE |
|---|---:|
| Training RMSE | 20.40 |
| Validation RMSE | 17.71 |

### Observations
- The model achieved relatively similar training and validation errors.
- Validation error being slightly lower than training error suggests the model is currently not overfitting.
- Since this is a purely linear model, its representational capacity is limited and may underfit complex environmental relationships.
- This baseline will be used as a reference point for later polynomial regression experiments.

### Next Steps
- Perform polynomial feature expansion (Degree 1 → 12)
- Compare training vs validation RMSE
- Visualize the bias–variance tradeoff
- Identify the degree where overfitting begins